File: Helpful_Code/mergeSort.cpp
#include <iostream>
#include <vector>

std::vector<int> merge(std::vector<int> mergeArr1, std::vector<int> mergeArr2) {
    std::vector<int> mergedArr;
    while (!mergeArr1.empty() && !mergeArr2.empty()) {
        if (mergeArr1[0] <= mergeArr2[0]) {
            mergedArr.push_back(mergeArr1[0]);
            mergeArr1.erase(mergeArr1.begin());
        } else if (mergeArr2[0] < mergeArr1[0]) {
            mergedArr.push_back(mergeArr2[0]);
            mergeArr2.erase(mergeArr2.begin());
        }
    }
    if (!mergeArr1.empty()) {
        mergedArr.insert(mergedArr.end(), mergeArr1.begin(), mergeArr1.end());
        mergeArr1.clear();
    }
    if (!mergeArr2.empty()) {
        mergedArr.insert(mergedArr.end(), mergeArr2.begin(), mergeArr2.end());
        mergeArr2.clear();
    }
    return mergedArr;
}

std::vector<int> mergeSort(std::vector<int> in_list) {
    if (in_list.size() <= 1) {
        return in_list;
    }
    int midpoint = in_list.size() / 2;
    std::vector<int> left(in_list.begin(), in_list.begin() + midpoint);
    std::vector<int> right(in_list.begin() + midpoint, in_list.end());
    left = mergeSort(left);
    right = mergeSort(right);
    return merge(left, right);
}

int main() {
    std::vector<int> exampleList = {4, 2, 8, 10, 42, 2, 4, 5, 90, 71, 17, 8};
    std::vector<int> sortedList = mergeSort(exampleList);
    for (int num : sortedList) {
        std::cout << num << " ";
    }
    std::cout << std::endl;
    return 0;
}

File: Helpful_Code/mergesort.py
def merge(mergeArr1: list[int], mergeArr2: list[int]) -> list[int]:
    mergedArr = []
    while len(mergeArr1) > 0 and len(mergeArr2) > 0:
        if mergeArr1[0] <=  mergeArr2[0]:
            mergedArr.append(mergeArr1[0])
            mergeArr1.pop(0)
            
        elif mergeArr2[0] < mergeArr1[0]:
            mergedArr.append(mergeArr2[0])
            mergeArr2.pop(0)
    if len(mergeArr1) > 0:
        mergedArr += mergeArr1
        mergeArr1 = []
    if len(mergeArr2) > 0:
        mergedArr += mergeArr2
        mergeArr2 = []
    return mergedArr

def mergeSort(in_list: list[int]):
    if len(in_list) <= 1:
        return in_list
    
    midpoint = len(in_list) // 2
    left = in_list[:midpoint]
    right = in_list[midpoint:]
    
    left = mergeSort(left)
    right = mergeSort(right)
    
    return merge(left, right)
    
def main():
    exampleList = [4, 2, 8, 10, 42, 2, 4, 5, 90, 71, 17, 8]
    print(mergeSort(exampleList))

if __name__ == "__main__":
    main()